library/std/src/thread/mod.rs

```diff
@@ -1739,7 +1739,11 @@ struct JoinInner<'scope, T> {
 impl<'scope, T> JoinInner<'scope, T> {
     fn join(mut self) -> Result<T> {
         self.native.join();
-        Arc::get_mut(&mut self.packet).unwrap().result.get_mut().take().unwrap()
+        if let Some(packet) = Arc::get_mut(&mut self.packet) {
+            packet.result.get_mut().take().unwrap()
+        } else {
+            Err(Box::new("thread terminated unexpectedly (e.g. due to OS intervention)"))
```

Comment:
Let's not speculate if we don't have anything else to go on:
```suggestion
            Err(Box::new("thread terminated unexpectedly"))
```

Reply:
Not sure if this is the right type to return.

Reply:
I do think it'd be better to return an `impl Error {}`  type that implements `Display` rather than just a `Box<str>`.

Reply:
Hmm. The `std::thread::Result` type states that its `Err` variant indicates that a thread has panicked with. So now it seems to me that I'm breaking the API, technically.

We should either use `.except(..)` instead, or change the documentation of `std::thread::Result`.